core/src/main/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProvider.scala

```diff
@@ -29,9 +31,34 @@ class LocalEntityProvider(workspace: Workspace, implicit protected val dataSourc
 
   private val workspaceContext = workspace
 
-  override def entityTypeMetadata(): Future[Map[String, EntityTypeMetadata]] = {
+  override def entityTypeMetadata(useCache: Boolean): Future[Map[String, EntityTypeMetadata]] = {
     dataSource.inTransaction { dataAccess =>
-      dataAccess.entityQuery.getEntityTypeMetadata(workspaceContext)
+      dataAccess.workspaceQuery.isEntityCacheCurrent(workspaceContext.workspaceIdAsUUID).flatMap { isCacheCurrentOpt =>
+        //If the cache doesn't exist; return full query results and mark the workspace for an update
+        if(isCacheCurrentOpt.isEmpty) {
+          dataAccess.workspaceQuery.updateCacheLastUpdated(workspaceContext.workspaceIdAsUUID, new Timestamp(workspaceContext.lastModified.getMillis - 1)).flatMap { _ =>
+            dataAccess.entityQuery.getEntityTypeMetadata(workspaceContext)
+          }
+        }
+        //Else , if the user wants to use the cache and it's up to date and it's enabled; return cached results
+        else if(useCache && cacheEnabled && isCacheCurrentOpt.getOrElse(false)) {
+          val typesAndCountsQ = dataAccess.entityTypeStatisticsQuery.getAll(workspaceContext.workspaceIdAsUUID)
+          val typesAndAttrsQ = dataAccess.entityAttributeStatisticsQuery.getAll(workspaceContext.workspaceIdAsUUID)
+
+          typesAndCountsQ flatMap { typesAndCounts =>
+            typesAndAttrsQ map { typesAndAttrs =>
+              (typesAndCounts.keySet ++ typesAndAttrs.keySet) map { entityType =>
+                (entityType, EntityTypeMetadata(
+                  typesAndCounts.getOrElse(entityType, 0),
+                  entityType + Attributable.entityIdAttributeSuffix,
+                  typesAndAttrs.getOrElse(entityType, Seq()).map(AttributeName.toDelimitedName).sortBy(_.toLowerCase)))
+              } toMap
+            }
+          }
```

Comment:
this block is copy-pasted from `EntityComponent` - could it be extracted into a reusable method that accepts `typesAndCountsQ`  and `typesAndAttrsQ` as arguments?